データベース設計

tamtam-time opened the pull request:
# What
データベース設計

# Why
アプリ作成の大本となる設計の確認のため

tamtam-time committed:
Revert "データベース設計"

This reverts commit c880d1e24f8210fe610baff1445d1ec0b66567fe.

ushi10 commented:
レビュー依頼ありがとうございます。
データベース設計はREADMEとER図両方を確認しレビューを行なっております。
詳細はissueに記載があるためそちらをご確認ください。

そのため、こちらのプルリクエストにER図をgyazoにて添付した後、再度レビュー依頼フォームをご提出ください。
どうぞよろしくお願いいたします。

tamtam-time commented:
失礼いたしました。
以下、ER図のgyazoになります。ご確認のほど宜しくお願いいたします。
https://gyazo.com/aab1e0e1e87bbff98c72f701801860f5

OhyaYouki reviewed (commented):
レビュー依頼ありがとうございます。
ER図の添付もありがとうございます！

お伝えしたいことが2点ございます。
1点目 **コードの修正がうまく差分として上がっていない状態です。**
こちらの原因は以前にコミットをリバートをされていることが挙げられます。
リバートされているコミットを今一度リバートするとこで解消されますので、ご対応のほどよろしくお願いいたします。
操作に関しまして、ご不明点などございましたら、通話またはチャットでご質問をお願いいたします。

2点目 **最終課題で必要になるテーブルは全部で4つとなります。**
（commentに関するテーブルは必須機能に含まれませんのでこちらは含めず4つです。）
user,itemに追加であと二つ必要になりますので、今一度見本アプリなど参考にテーブル設計に挑戦して見ましょう。


コードの修正後、コミット・プッシュを行った上で再度レビュー依頼をお送りください。
ご対応のほど、よろしくお願いいたします。

tamtam-time committed:
データベース設計：修正分

以下、ER図のURLになります
https://gyazo.com/01aced6369f1a976c81d27ad5edbc911

tamtam-time commented:
ER図・READ MEの修正分になります。ご確認のほど宜しくお願いいたします。

minachanchan commented on README.md:
```diff
@@ -10,20 +12,12 @@
 | first_name         |string  | null: false |
 | last_name          |string  | null: false |
 | birthday           |integer | null: false |
```
```ruby
| email |string | null: false |
```
こちらでのコメントで失礼いたします。

emailは重複を避けるため、オプションにunique: trueをつけましょう！

minachanchan replied:
```ruby
| first_name         |string  | null: false |
| last_name          |string  | null: false |
```

ユーザーの本名と、フリガナに関しては以下のように4つの項目に分け、
それぞれのカラムを設定するよう修正しましょう！
```
・姓
・名
・姓のフリガナ
・名のフリガナ
```

minachanchan commented on README.md:
```diff
@@ -10,20 +12,12 @@
 - has_many :items
-- has_many :comments
+- has_many :buys
+- has_one :cards
```
```ruby
- has_many :buys
- has_one :cards
```
ユーザーと紐づくテーブルは商品情報テーブルと購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■ユーザーと商品情報（18行目で指定済み）
ユーザー：親
商品情報：子
1対多で指定

■ユーザーと購入履歴
ユーザー：親
購入履歴：子
1対多で指定
```

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
 | area       | string     | null: false                     |
 | days       | integer    | null: false                     |
```
```ruby
| category   | string     | null: false                     |
| status     | string     | null: false                     |
| postage    | integer    | null: false                     |
| area       | string     | null: false                     |
| days       | integer    | null: false                     |
```
こちらはプルダウンで選択する項目とお見受けします。
active_hashでの実装に際し、以下についてご確認くださいませ。

①
active_hashを使用して保存するデータは数値を扱うので、
データ型はintegerを指定しましょう！

②
active_hashを使用して保存するデータのカラムは「単数系_id」という形にする必要がございます。
以下のように各カラム名を修正しましょう！

```
例：category_id
```

③
「day」は予約語ですので、使用すると予期せぬエラーが起こる恐れがあります。
「day」以外の単語を使用したカラム名に変更しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
 | area       | string     | null: false                     |
 | days       | integer    | null: false                     |
 | price      | integer    | null: false                     |
```
どのユーザーが出品者なのかわかるよう、出品者のidを保存するカラムが必要です。
ユーザーテーブルの外部キーカラムを追加しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+### Association
+
+- belongs_to :users
+- has_one :buys
```
```ruby
- belongs_to :users
- has_one :buys
```
商品情報と紐づくテーブルはユーザーテーブルと購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■商品情報とユーザー
ユーザー：親
商品情報：子
1対多で指定

■商品情報と購入履歴
商品情報：親
購入履歴：子
1対1で指定
```

また、belongs_to、has_oneでの指定なので、テーブル名は単数形で指定しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+- belongs_to :users
+- has_one :buys
+
+## buys テーブル
```
記載いただいているカラムから、一度こちらを配送先住所情報テーブルと認識し、コメントいたします。
配送先住所情報に関するテーブルとしてテーブル名を変更することをお勧めいたします。

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+
+| Column             | Type       | Options                         |
+| ------------------ | ---------- | --------------------------------|
+| postalcode         |integer     | null: false                     |
```
```ruby
| postalcode |integer | null: false |
```

郵便番号には「-」を含める必要があるため、string型を指定しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+| Column             | Type       | Options                         |
+| ------------------ | ---------- | --------------------------------|
+| postalcode         |integer     | null: false                     |
+| prefectures        |string      | null: false                     |
```
```ruby
| prefectures |string | null: false |
```
都道府県を保存するカラムでは、商品情報テーブルで使用したactive_hashのデータを兼用できます。
商品情報で指定したものと同じものをこちらでも指定しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+| municipaldistrict  |string      | null: false                     |
+| district           |string      | null: false                     |
+| building           |string      |                                 |
+| phone              |integer     | null: false                     |
```
```ruby
| phone |integer | null: false |
```
電話番号の多くは0から始まりますが、integer型は0から始まる整数を識別できません。
そのため、電話番号はstring型で指定しましょう！

minachanchan replied:
購入履歴テーブルの外部キーカラムを追加しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
 
-- has_many :comments
+- has_many :users
+- has_one :cards
```
```ruby
- has_many :users
- has_one :cards
```

発送先住所と紐づくテーブルは購入履歴テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■発送先住所と購入履歴
購入履歴：親
発送先住所：子
1対1
```

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+- has_one :cards
 
-## comments テーブル
+## cards テーブル
```
今回のフリマアプリではカード情報を直接テーブルには保存しません。
商品購入履歴テーブルが必要になりますのでそちらを実装しましょう。
購入履歴テーブルとしてテーブル名を変更することをお勧めいたします。

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+| ------------------ | ------ | ----------- |
+| card_number        |integer | null: false |
+| deadline           |integer | null: false |
+| code               |integer | null: false |
```
```ruby
| card_number        |integer | null: false |
| deadline           |integer | null: false |
| code               |integer | null: false |
```
カード情報は保存しませんのでこちらは削除しましょう。
代わりに「誰が」「どの商品を」購入したか管理できるように、ユーザー情報テーブルの外部キーカラム、商品情報テーブルの外部キーカラムを記述しましょう！

minachanchan commented on README.md:
```diff
@@ -37,18 +31,35 @@
+| code               |integer | null: false |
 
+- belongs_to :users
+- belongs_to :buys
```
```ruby
- belongs_to :users
- belongs_to :buys
```
購入履歴と紐づくテーブルはユーザーテーブル、商品情報テーブル、発送先住所テーブルです。
以下の親子関係で、適切なアソシエーションを指定しましょう！

```
■購入履歴とユーザー
ユーザー：親
購入履歴：子
1対多で指定

■購入履歴と商品情報
商品情報：親
購入履歴：子
1対1

■購入履歴と発送先住所
購入履歴：親
発送先住所：子
1対1
```

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビューのご依頼頂けますようお願い申し上げます。
その際、**ER図もREADMEの内容と揃えていただき、次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

何卒よろしくお願いいたします。

tamtam-time committed:
データベース設計：修正分

以下、ER図のURLになります。
https://gyazo.com/1ee61decaecb2e125a1e175b8ea8c575

アプリ見本では購入履歴の画面が確認できず、また、購入したのは勿論ログインしている本人であるのに購入者の情報が必要という点で、不明点が多く、また誤りが多数あるかと思います。申し訳ございませんが、何卒宜しくお願いいたします。

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
-- has_many :comments
+- belongs_to :items
+- belongs_to :destinations
+- belongs_to :historys
```
```ruby
- belongs_to :items
- belongs_to :destinations
- belongs_to :historys
```
ユーザー情報テーブルで組むべきアソシエーションは以下二つです。
改めてご精査お願いいたします。
```
・複数の商品情報と紐づく
・複数の購入履歴と紐づく
```

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
-| days       | integer    | null: false                     |
+| postage_id | integer    | null: false                     |
+| area_id    | integer    | null: false                     |
+| send_day   | integer    | null: false                     |
```
```ruby
| category   | string     | null: false                     |
| status     | string     | null: false                     |

| send_day | integer | null: false |
```
こちらについても以下コメント①、②の修正をお願いいたします。
https://github.com/tamtam-time/furima-38599/pull/10#discussion_r1065320803

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
+### Association
+
+- has_one :user
+- belongs_to :history
```
```ruby
- has_one :user
- belongs_to :history
```
以下の関係性に修正しましょう。
```
・一つのユーザー情報に所属する
・一つの購入履歴と紐づく
```

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
+| Column             | Type       | Options                         |
+| ------------------ | ---------- | --------------------------------|
+| postalcode         |string      | null: false                     |
+| prefectures_id     |string      | null: false                     |
```
```ruby
| prefectures_id |string | null: false |
```
https://github.com/tamtam-time/furima-38599/pull/10#discussion_r1065322283
こちらについて再度ご確認をお願いいたします。
商品情報テーブルの都道府県カラムと同じものを指定しましょう。

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
 
-- has_many :comments
+- has_many :users
+- belongs_to :historys
```
```ruby
- has_many :users
- belongs_to :historys
```
https://github.com/tamtam-time/furima-38599/pull/10#discussion_r1065322909
こちらご参照の上適切なアソシエーションに修正しましょう。

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
+
+| Column             | Type        | Options                         |
+| ------------------ | ----------- | ------------------------------- |
+| item_name          | string      | null: false                     |
```
```ruby
| item_name | string | null: false |
```
こちらのカラムは不要です。
削除しましょう。

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
+| ------------------ | ----------- | ------------------------------- |
+| item_name          | string      | null: false                     |
+| user               | references  | null: false, foreign_key: true  |
+| destination        | references  | null: false, foreign_key: true  |
```
```ruby
| destination | references | null: false, foreign_key: true |
```
https://github.com/tamtam-time/furima-38599/pull/10#discussion_r1065323848
こちらでご案内の通り、商品情報テーブルの外部キーカラムに修正しましょう。

minachanchan commented on README.md:
```diff
@@ -1,54 +1,77 @@
+- belongs_to :history
 
+- has_one :destination
+- belongs_to :history
```
```ruby
- has_many :users
- belongs_to :historys

- has_one :item
- belongs_to :history

- has_one :destination
- belongs_to :history
```
https://github.com/tamtam-time/furima-38599/pull/10#discussion_r1065324057
こちら参照の上再度ご精査をお願いいたします。
その上で不要なものは削除しましょう。

minachanchan reviewed (commented):
実装お疲れ様です！レビューのご依頼ありがとうございます。

コメントいたしましたので、ご確認頂けますと幸いでございます。

指摘内容の修正がお済みになりましたら、再度レビューのご依頼頂けますようお願い申し上げます。
その際、**ER図もREADMEの内容と揃えていただき、次回のレビュー依頼の際に最新のER図のURLをまたこちらのプルリクエストに添付頂ければと存じます。**

ご不明点がございましたらコネクトにてご質問くださいませ。
何卒よろしくお願いいたします。

tamtam-time committed:
データベース設計：修正分

以下、ER図のURLになります
https://gyazo.com/87671e9bc298ff953e5bc257a3333c0a

ご確認のほど宜しくお願いいたします。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,71 @@
+| first_name_yomi    |string      | null: false                     |
+| last_name_yomi     |string      | null: false                     |
+| birthday           |integer     | null: false                     |
+| history            |references  | null: false, foreign_key: true  |
```
```
| history            |references  | null: false, foreign_key: true  |
```
購入管理テーブルとユーザー管理テーブルは親子関係ではありません。
このため、ユーザーテーブルに購入管理テーブルの外部キーは不要です。
削除しておきましょう。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,71 @@
 - has_many :items
-- has_many :comments
+- belongs_to :historys
 
```
```
- belongs_to :historys
```
ユーザーはいくつのもの商品を購入することができます。
上記を踏まえて、アソシエーションの記述を修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,71 @@
-| user      | references | null: false, foreign_key: true |
+### Association
 
+- belongs_to :users
```
```
- belongs_to :users
```
アソシエーションにbelongs_toを用いる際はモデル名は単数形となります。
修正しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,71 @@
+| Column             | Type       | Options                         |
+| ------------------ | ---------- | --------------------------------|
+| postalcode         |string      | null: false                     |
+| prefectures_id     |integer     | null: false                     |
```
```
| prefectures_id     |integer     | null: false                     |
```

Active_hashのデータは使い回すことができますので、itemテーブルの都道府県カラムと同じものを使用しましょう。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,71 @@
+| last_name          |string      | null: false                     |
+| first_name_yomi    |string      | null: false                     |
+| last_name_yomi     |string      | null: false                     |
+| birthday           |integer     | null: false                     |
```
```
| birthday           |integer     | null: false                     |
```
今回の実装では、誕生日カラムは「date型」に修正しましょう。（date型は年、月、日を1つで指定して取得できるため）

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。
また、最新のER図の添付をお願いいたします。

tamtam-time committed:
データベース設計：修正分

以下が最新のER図になります。
https://gyazo.com/0294c5662f1475d13204c513db7b206a

ご確認のほど宜しくお願いいたします。

maimai0401 commented on README.md:
```diff
@@ -1,54 +1,70 @@
 ### Association
 
-- has_many :comments
+- has_many :users
```
```
- has_many :users
```
配送先住所テーブルは購入管理テーブルと結びついているため、ユーザーテーブルとのアソシエーションは不要です。
削除しておきましょう。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
修正点確認させていただきました。
コメントいたしましたので、ご確認お願いいたします。
修正後、再度レビュー依頼フォームにてご提出くださいませ。

また、最新のER図の添付をお願いいたします。

tamtam-time committed:
データベース設計：修正分

以下、最新のER図になります。
https://gyazo.com/ba412c30d557da08d595c537792fe390

ご確認のほど宜しくお願いいたします。

maimai0401 reviewed (commented):
レビュー依頼ありがとうございます。
LGTMとさせていただきます！実装お疲れ様でした。

マージ後、ローカルのmasterブランチに忘れずにプルしましょう。
その後、新しいブランチを作成し次の実装に取り組んでいただけますと幸いです。
引き続き実装頑張っていきましょう！
